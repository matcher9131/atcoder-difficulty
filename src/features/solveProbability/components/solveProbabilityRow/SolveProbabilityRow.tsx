import type { ReactNode } from "react";

export type SolveProbabilityRowProps = {
    readonly contestHeaderCell: ReactNode;
    readonly problemCell: ReactNode;
    readonly difficulty: string;
    readonly solveProbability: string;
};

export const SolveProbabilityRow = ({
    contestHeaderCell,
    problemCell,
    difficulty,
    solveProbability,
}: SolveProbabilityRowProps): ReactNode => {
    return (
        <tr className="contents">
            {contestHeaderCell}
            {problemCell}
            <td className="px-3 py-6 text-right">{difficulty}</td>
            <td className="px-3 py-6 text-right">{solveProbability}</td>
        </tr>
    );
};
