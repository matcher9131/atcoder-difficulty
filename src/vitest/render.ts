import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
