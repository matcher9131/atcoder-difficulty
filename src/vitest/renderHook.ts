import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
