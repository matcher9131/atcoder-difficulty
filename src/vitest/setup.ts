import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
