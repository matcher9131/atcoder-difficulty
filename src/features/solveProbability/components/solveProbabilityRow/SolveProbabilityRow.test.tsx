import { describe, it, expect, vi, beforeEach } from "vitest";
import { screen, fireEvent } from "@testing-library/react";
import { SolveProbabilityRow, type SolveProbabilityRowProps } from "./SolveProbabilityRow";
import { render } from "../../../../vitest/render";

// Mock useTranslation
vi.mock("react-i18next", () => ({
    useTranslation: () => ({
        t: (key: string) => key,
    }),
}));

const contestHeaderCell = <td data-testid="contest-header">Contest</td>;
const problemCell = <td data-testid="problem-cell">Problem</td>;

const defaultProps: SolveProbabilityRowProps = {
    contestHeaderCell,
    problemCell,
    difficulty: "1234",
    solveProbability: "56%",
    onGraphButtonClick: vi.fn(),
};

describe("SolveProbabilityRow", () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("renders all cells in correct order", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        const cells = screen.getAllByRole("cell");
        expect(cells[0]).toContainElement(screen.getByTestId("contest-header"));
        expect(cells[1]).toContainElement(screen.getByTestId("problem-cell"));
        expect(cells[2]).toHaveTextContent(defaultProps.difficulty);
        expect(cells[3]).toHaveTextContent(defaultProps.solveProbability);
        expect(cells[4].querySelector("button")).toBeInTheDocument();
    });

    it("renders difficulty and solveProbability", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        expect(screen.getByText(defaultProps.difficulty)).toBeInTheDocument();
        expect(screen.getByText(defaultProps.solveProbability)).toBeInTheDocument();
    });

    it("renders contestHeaderCell and problemCell", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        expect(screen.getByTestId("contest-header")).toBeInTheDocument();
        expect(screen.getByTestId("problem-cell")).toBeInTheDocument();
    });

    it("renders button with correct label and class", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        const button = screen.getByRole("button");
        expect(button).toHaveTextContent("solveProbabilityTableContent.showGraphButtonLabel");
        expect(button).toHaveClass("btn");
    });

    it("calls onGraphButtonClick when button is clicked", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        expect(defaultProps.onGraphButtonClick).toHaveBeenCalledTimes(1);
    });

    it("applies correct cell class names", () => {
        render(<SolveProbabilityRow {...defaultProps} />);
        const cells = screen.getAllByRole("cell");
        expect(cells[2]).toHaveClass("justify-end");
        expect(cells[3]).toHaveClass("justify-end");
        expect(cells[4]).toHaveClass("justify-center");
    });

    it("matches snapshot", () => {
        const { container } = render(<SolveProbabilityRow {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });
});
