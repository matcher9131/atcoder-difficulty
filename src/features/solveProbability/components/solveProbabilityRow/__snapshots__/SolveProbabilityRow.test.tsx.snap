// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SolveProbabilityRow > matches snapshot 1`] = `
<div>
  <tr
    class="contents"
  >
    <td
      data-testid="contest-header"
    >
      Contest
    </td>
    <td
      data-testid="problem-cell"
    >
      Problem
    </td>
    <td
      class="flex px-3 py-auto h-18 items-center border-t-1 border-base-content/75 group justify-end"
    >
      1234
    </td>
    <td
      class="flex px-3 py-auto h-18 items-center border-t-1 border-base-content/75 group justify-end"
    >
      56%
    </td>
    <td
      class="flex px-3 py-auto h-18 items-center border-t-1 border-base-content/75 group justify-center"
    >
      <button
        class="btn"
      >
        solveProbabilityTableContent.showGraphButtonLabel
      </button>
    </td>
  </tr>
</div>
`;
